Feat: Login UI Issue#14

mtreacy002 opened the pull request:
### Description
Add Login UI

Fixes #14 and #29 

### Type of Change:

- Code
- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
- running npm start and submit without input. Form validation `required` gave warning as expected.
- login with existing user but with wrong password or username, error message is shown as expected.
- login with correct credentials, user redirected to Member Portfolio page as expected.

[link to Gif](https://media.giphy.com/media/IfxCozlQxlr1xwroA5/giphy.gif)

![giphy showing login screen](https://user-images.githubusercontent.com/11148726/86510952-5ab58b00-bdec-11ea-9655-afb8d4b0ee5f.gif)

Note to tester/reviewer. To test this login ui you must run the backend server from the latest code change on login backend [PR#60](https://github.com/anitab-org/bridge-in-tech-backend/pull/60)

### Checklist:

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials


**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings

**Additional Note**
Up to User Interface only, backend API consumption is not yet added - wip

mtreacy002 committed:
Feat: Allow user to Login issue#14

mtreacy002 commented:
Update @anitab-org/bridgeintech-maintainers , Just pushed this Login form as user interface. Form validation `required` has been applied but backend API consumption is WIP.

mtreacy002 commented:
Update @anitab-org/bridgeintech-maintainers . This form now has error responses shown to user. I still need to workout how to persist the jwt token that got returned and carry that to the next page. @meenakshi-dhanani if you can give me some guidance in this, that'll be great. 😉

meenakshi-dhanani commented:
Try and have more meaningful commit messages which indicate intent. eg. Allow a user to login.

meenakshi-dhanani commented on src/Routes.js:
```diff
@@ -33,6 +35,7 @@ export default (
       <Route exact path="/" component={Home} />
         <Route path="/register" component={Register} />
         <Route path="/login" component={Login} />
+        <Route path="/my_space" component={MySpace} />
```
I wasn't sure about this. But maybe a hyphen(-) is preferred over underscore(_). This has some details - [Hypher over Underscore](https://www.ecreativeim.com/blog/index.php/2011/03/30/seo-basics-hyphen-or-underscore-for-seo-urls/#:~:text=The%20short%20answer%20is%20that,treat%20an%20underscore%20that%20way.&text=This%20has%20been%20confirmed%20directly,a%20(minor)%20ranking%20benefit.)

mtreacy002 replied:
done. changed it to hyphen

meenakshi-dhanani commented on src/login/Login.css:
```diff
@@ -0,0 +1,52 @@
+
+
+.login-form {
+width: 80%;
```
formatting? you might need to indent here

mtreacy002 replied:
fixed

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+                    <form className="login-form mx-auto" onSubmit={handleSubmit}>
+                        <form-group controlId="formUserame">
+                            <p className="input-control">
+                                <label htmlFor="username">Username or Email:</label>
```
remove the or email. I might expect login via email too. It can be misleading at this point, we can look at it later

mtreacy002 replied:
The functionality is actually already there for login with with email. I just don't have test to check successful login done with email (coz they are just mocking anyway). So, I'll leave the option there (Username or Email)

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+
+
+export default function Login() {
+    const [username, setUsername] = useState(true);
```
I see, the value of username, password, accessToken and accessExpiry is never read. Then why are you setting it? you don't need it.

mtreacy002 replied:
does that mean I should remove onChange on each field attributes? Or what should I replace it with? How will the field changes get recorded?

<img width="543" alt="Screen Shot 2020-06-25 at 9 46 35 am" src="https://user-images.githubusercontent.com/29667122/85638320-efbbd400-b6c8-11ea-8033-caf371f38cf6.png">

mtreacy002 replied:
Done, remove onChange

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+
+        new FormData(e.target).forEach((value, key) => {
+            if (key === "username")
+                userInput[key] = value;
```
Do you really need an if else statement here? The action that you are performing is the same either ways. 
Why can't it simply be userInput[key] = value?

mtreacy002 replied:
Yep, fixed

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+
+
+        const payload = {
+            username: userInput["username"],
```
userInput has a username and password that you are assigning to payload. Can you not use userInput, why make a copy of payload.

mtreacy002 replied:
fixed

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+                                    type="text"
+                                    name="username"
+                                    placeholder="Username or Email"
+                                    onChange={e => setUsername(e.target.value)}
```
there is no need to set here. You are fetching the values from FormData anyway. So even if you set here, it's not being used

mtreacy002 replied:
done. remove onChange

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+                        </form-group>
+                        <div><br></br></div>
+                        <div>
+                            {responseMessage !== null && <span className="error" name="response" aria-label="response" role="alert">{responseMessage}</span>}
```
since this responseMessage will only show up in case there is an error, can we call this errorMessage?

mtreacy002 replied:
done

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,133 @@
+        fetch("http://127.0.0.1:5000/login", requestLogin)
+            .then(async response => {
+                resStatus = response.status
+                data = await response.json();
```
This can be refactored. A switch is not required. I can catchup with you to help refactor.

mtreacy002 replied:
yep, let's catch up tonight same time?

meenakshi-dhanani replied:
I will be available at 7pm IST today. would that work for you?

meenakshi-dhanani commented on src/tests/Login.test.jsx:
```diff
@@ -0,0 +1,78 @@
+            target: { value: 'true' },
+        })
+    });
+    expect(screen.queryByLabelText("response")).not.toBeInTheDocument();
```
would have been better if the label was errorMessage. So the test would make sense and it would feel login is successful if there is no errorMessage in the document.

mtreacy002 replied:
done

mtreacy002 commented:
Update @anitab-org/bridgeintech-maintainers and @meenakshi-dhanani . I've just pushed the changes requested in your review. Can you please re-review? Thanks

mtreacy002 commented:
Posted this in a wrong PR 🤣

> update @anitab-org/bridgeintech-maintainers . Successful logged in user is now redirected to Member Portfolio page.

[link to GIF](https://media.giphy.com/media/IfxCozlQxlr1xwroA5/giphy.gif)

mtreacy002 commented:
@meenakshi-dhanani and @anitab-org/bridgeintech-maintainers . I've pull rebased the latest merged to develop and fixed few things mentioned by Meena on the feedback;
* changed switch to if else, 
* removed unused useState (access token and access expiry)
* redo setting payload logic

PS: While I'm on it, I did the same in Register.js (changed setting up the payload logic and removed unused useState) for consistency.

meenakshi-dhanani commented on src/login/Login.jsx:
```diff
@@ -0,0 +1,113 @@
+
+
+export default function Login() {
+    const [errorMessage, setResponseMessage] = useState(null);
```
errorMessage, setErrorMessage - try following the convention

mtreacy002 replied:
yep, missed that 😁

meenakshi-dhanani commented on src/tests/Login.test.jsx:
```diff
@@ -0,0 +1,78 @@
+
+
+
+it('allows the user to login successfully', async () => {
```
We might want to have another test that checks if on successful login, redirect to My Space happens. That test might be tricky, so we can have a separate PR for that, not on priority though, so as not to block you

mtreacy002 committed:
PR#26 Handle server error and refactor fetch

meenakshi-dhanani reviewed (approved):
I have tested the following scenarios:

 - Redirect to My Space on successful login
 - Show error message when server is down
 - Show appropriate error for invalid credentials (401)

The unit tests for these also have passed. Approving!

meenakshi-dhanani commented:
@foongminwong could you test these? I've done a code review for Maya. I'm hoping you can add gifs and screenshots 😄

foongminwong reviewed (approved):
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done by @meenakshi-dhanani 

2. All possible responses were tested as below (tap/zoom in to see clearer gifs):

| **Login API**  |  **Login UI**  | **Expected Result** | **Actual Result** |
|---|---|---|---|
| ![bit-login-wrong-password](https://user-images.githubusercontent.com/23441795/85912487-7b7d4f80-b7f2-11ea-860e-cef2560fddb1.gif) | ![bit-login-wrong-password-1](https://user-images.githubusercontent.com/23441795/85912489-7ddfa980-b7f2-11ea-84f4-da77cdf070f9.gif) |  As a user, when I login with incorrect username/password, it should show an error.  |  ✅ |
 | ![bit-login-successful](https://user-images.githubusercontent.com/23441795/85912552-c9925300-b7f2-11ea-901e-deea1b2950f3.gif)  |  ![bit-login-successful-1](https://user-images.githubusercontent.com/23441795/85912556-cdbe7080-b7f2-11ea-85e3-30c745f01f3d.gif) | As a user, when I login with correct credentials, it should direct me to a BIT success landing page.  | ✅ |
| ![bit-login-user-pass-field-missing](https://user-images.githubusercontent.com/23441795/85912573-052d1d00-b7f3-11ea-8122-721a70b9d0a3.gif) | ![bit-login-user-pass-field-missing-1](https://user-images.githubusercontent.com/23441795/85912574-078f7700-b7f3-11ea-84d5-7599378830b1.gif) | As a user, if I leave one of the fields blank, it should prompt me to enter values for username/password.  | ✅  |
|  ![bit-login-unverify-email](https://user-images.githubusercontent.com/23441795/85912591-3a396f80-b7f3-11ea-9c8e-9e63312e53e0.gif) | ![bit-login-unverify-email-1](https://user-images.githubusercontent.com/23441795/85912593-3c033300-b7f3-11ea-920f-efdceb60ef7f.gif) | As a user, when I login with an unverified account, it should tell me verify my account first before logging in.|  ✅  |

    
3. Additional testcases covered: N/A
    
4. Additional Comments: 
* Great job @mtreacy002 !! Merge Time ⏲

5. OS Version: Windows 10

mtreacy002 commented:
Wowww....That's an amazing tests report, @foongminwong. Thank youuu... 😍
